File: src/joint_embedding/unit_tests/test_method.R
library(testthat, quietly = TRUE)
requireNamespace("anndata", quietly = TRUE)

## VIASH START
# This code block will be replaced by viash at runtime.
meta <- list(functionality_name = "foo")
## VIASH END

method_id <- meta[["functionality_name"]]
command <- paste0("./", method_id)

# define some filenames
testpar <- list(
  input_mod1 = "resources_test/joint_embedding/test_resource.mod1.h5ad",
  input_mod2 = "resources_test/joint_embedding/test_resource.mod2.h5ad",
  output = "output.h5ad"
)

cat("> Running method\n")
out <- processx::run(
  command = paste0("./", meta[["functionality_name"]]),
  args = c(
    "--input_mod1", testpar$input_mod1,
    "--input_mod2", testpar$input_mod2,
    "--output", testpar$output
  ),
  stderr_to_stdout = TRUE
)

cat("> Checking whether output files were created\n")
expect_true(file.exists(testpar$output))

cat("> Reading h5ad files\n")
input_mod1 <- anndata::read_h5ad(testpar$input_mod1)
output <- anndata::read_h5ad(testpar$output)

cat("> Checking contents of output.h5ad\n")
expect_equal(output$uns[["dataset_id"]], input_mod1$uns[["dataset_id"]])
expect_equal(output$uns[["method_id"]], method_id)
expect_equal(output$n_obs, input_mod1$n_obs)
expect_gte(output$n_vars, 1)
expect_lte(output$n_vars, 100)
expect_false(is.null(output$obs_names))
expect_equal(output$obs_names, input_mod1$obs_names)

cat("> Test succeeded!\n")

File: src/joint_embedding/unit_tests/test_method.py
from os import path
import subprocess
import anndata as ad

par = {
    "input_mod1": "resources_test/joint_embedding/test_resource.mod1.h5ad",
    "input_mod2": "resources_test/joint_embedding/test_resource.mod2.h5ad",
    "output": "output.h5ad",
}

print("> Running method")
out = subprocess.check_output([
  "./" + meta['functionality_name'],
  "--input_mod1", par['input_mod1'],
  "--input_mod2", par['input_mod2'],
  "--output", par['output']
]).decode("utf-8")

print("> Checking whether output files were created")
assert path.exists(par['output'])

print("> Reading h5ad files")
input_mod1 = ad.read_h5ad(par['input_mod1'])
output = ad.read_h5ad(par['output'])

print("> Checking contents of output.h5ad")
assert output.uns['dataset_id'] == input_mod1.uns['dataset_id']
assert output.uns['method_id'] == meta['functionality_name']
assert output.n_obs == input_mod1.n_obs
assert output.n_vars >= 1
assert output.n_vars <= 100
assert all(output.obs_names == input_mod1.obs_names)

print("> Test succeeded!")